httpclient5/src/main/java/org/apache/hc/client5/http/impl/async/ContentCompressionAsyncExec.java

```diff
@@ -0,0 +1,212 @@
+     *                      otherwise a {@link HttpException} is thrown.
+     */
+    public ContentCompressionAsyncExec(
+            final LinkedHashMap<String, UnaryOperator<AsyncEntityConsumer<?>>> decoderMap,
+            final boolean ignoreUnknown) {
+
+        Args.notEmpty(decoderMap, "Decoder map");
+
+        this.acceptEncoding = MessageSupport.headerOfTokens(
+                HttpHeaders.ACCEPT_ENCODING, new ArrayList<>(decoderMap.keySet()));
+
+        final RegistryBuilder<UnaryOperator<AsyncEntityConsumer<?>>> rb = RegistryBuilder.create();
+        decoderMap.forEach(rb::register);
+        this.decoders = rb.build();
+        this.ignoreUnknown = ignoreUnknown;
+    }
+
+    /**
+     * Convenience constructor equivalent to
+     * {@code new ContentCompressionAsyncExec(true)}.
+     */
+    public ContentCompressionAsyncExec() {
+        final LinkedHashMap<String, UnaryOperator<AsyncEntityConsumer<?>>> defaultMap =
+                new LinkedHashMap<>();
+
+        defaultMap.put(ContentCoding.DEFLATE.token(),
+                d -> new InflatingAsyncEntityConsumer<>(d, null));
+
+        this.acceptEncoding = MessageSupport.headerOfTokens(
+                HttpHeaders.ACCEPT_ENCODING, Collections.singletonList(ContentCoding.DEFLATE.token()));
+
+        this.decoders = RegistryBuilder.<UnaryOperator<AsyncEntityConsumer<?>>>create()
+                .register(ContentCoding.DEFLATE.token(),
+                        defaultMap.get(ContentCoding.DEFLATE.token()))
+                .build();
+        this.ignoreUnknown = true;
+    }
+
+
+    @Override
+    public void execute(final HttpRequest request,
+                        final AsyncEntityProducer producer,
+                        final AsyncExecChain.Scope scope,
+                        final AsyncExecChain chain,
+                        final AsyncExecCallback cb)
+            throws IOException, HttpException {
+
+        final HttpClientContext ctx = scope != null ? scope.clientContext : HttpClientContext.create();
+        final boolean enabled = ctx.getRequestConfigOrDefault().isContentCompressionEnabled();
+
+        if (enabled && !request.containsHeader(HttpHeaders.ACCEPT_ENCODING)) {
+            request.addHeader(acceptEncoding);
+        }
+
+        chain.proceed(request, producer, scope, new AsyncExecCallback() {
+
+            @Override
+            public AsyncDataConsumer handleResponse(final HttpResponse rsp,
+                                                    final EntityDetails details)
+                    throws HttpException, IOException {
+
+                if (!enabled) {
+                    return cb.handleResponse(rsp, details);
+                }
+
+                final String raw = details != null && details.getContentEncoding() != null
+                        ? details.getContentEncoding()
+                        : Optional.ofNullable(rsp.getFirstHeader(HttpHeaders.CONTENT_ENCODING))
```

Comment:
@arturobernalg This is unnecessary and can potentially produce unexpected results. The interceptor should rely on the properties of `EntityDetails` only.